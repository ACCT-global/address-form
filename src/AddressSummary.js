import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShape from './propTypes/AddressShape'
import { getField } from './selectors/fields'
import { POSTAL_CODE } from './constants'

class AddressSummary extends Component {
  render() {
    if (this.props.giftRegistryDescription) {
      return (
        <span>
          <span>At address of:</span>{' '}
          <strong className="gift-list-name">
            {this.props.giftRegistryDescription}
          </strong>
        </span>
      )
    }

    const { rules, address, canEditData, children } = this.props
    const postalCodeByInput = rules.postalCodeFrom === POSTAL_CODE
    const numberField = getField('number', rules)
    const complementField = getField('complement', rules)
    const neighborhoodField = getField('neighborhood', rules)

    const {
      street,
      country,
      number,
      complement,
      neighborhood,
      city,
      state,
      postalCode,
    } = address

    return (
      <div className="address-summary">
        {street && <span className="street">{street}</span>}

        {numberField && number && <span className="number-comma">{', '}</span>}
        {numberField && number && <span className="number">{number}</span>}

        {!canEditData && ' '}
        {!canEditData && (
          <a
            data-i18n="[title]modal.maskedInfoHello"
            className="client-masked-info"
            onClick={this.props.onClickMaskedInfoIcon}
          >
            <i className="icon-question-sign" />
          </a>
        )}

        {street || number || complement ? <br /> : null}

        {complementField &&
          complement && <span className="complement">{complement}</span>}

        {neighborhood &&
          complement && <span className="neighborhood-comma">{', '}</span>}
        {neighborhoodField &&
          neighborhood && <span className="neighborhood">{neighborhood}</span>}

        {city && <span className="city-dash">{' - '}</span>}
        {city && <span className="city">{city}</span>}

        {state && <span className="state-dash">{' - '}</span>}
        {state && <span className="state">{state}</span>}

        {postalCode || country ? <br /> : null}
        {postalCodeByInput && <span className="postal-code">{postalCode}</span>}

        {country && <span className="country-dash">{' - '}</span>}
        {country && <span className="country">{country}</span>}

        {children}
      </div>
    )
  }
}

AddressSummary.defaultProps = {
  canEditData: true,
}

AddressSummary.propTypes = {
  canEditData: PropTypes.bool,
  address: AddressShape.isRequired,
  rules: PropTypes.object.isRequired,
  children: PropTypes.node,
  giftRegistryDescription: PropTypes.string,
  onClickMaskedInfoIcon: PropTypes.func,
}

export default AddressSummary
