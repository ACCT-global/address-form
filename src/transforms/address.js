import reduce from 'lodash/reduce'
import filter from 'lodash/filter'
import map from 'lodash/map'
import difference from 'lodash/difference'
import find from 'lodash/find'
import findIndex from 'lodash/findIndex'
import { getField } from '../selectors/fields'
import { validateChangedFields } from '../validateAddress'
import msk from 'msk'

export function addValidation(address) {
  return reduce(
    address,
    (newAddress, value, propName) => {
      newAddress[propName] = { value }
      return newAddress
    },
    {}
  )
}

export function removeValidation(address) {
  return reduce(
    address,
    (newAddress, { value }, propName) => {
      newAddress[propName] = value
      return newAddress
    },
    {}
  )
}

export function addNewField(address, fieldName, value) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = {
        ...prop,
        [fieldName]: value,
      }
      return newAddress
    },
    {}
  )
}

export function removeField(address, fieldName) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = { ...prop }
      delete newAddress[propName][fieldName]
      return newAddress
    },
    {}
  )
}

export function addDisabledToProtectedFields(fields, rules) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasValue = prop && prop.value
      const isProtectField =
        rules.postalCodeProtectedFields &&
        rules.postalCodeProtectedFields.indexOf(propName) !== -1

      newFields[propName] = prop

      if (isProtectField && hasValue) {
        newFields[propName] = {
          ...prop,
          disabled: true,
        }
      }

      return newFields
    },
    {}
  )
}

export function handleMultipleValues(fields) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasMultipleValues = prop.value.indexOf(';') !== -1

      newFields[propName] = prop

      if (hasMultipleValues) {
        newFields[propName] = {
          ...prop,
          value: null,
          valueOptions: prop.value.split(';'),
        }
      }

      return newFields
    },
    {}
  )
}

export function maskFields(addressFields, rules) {
  return reduce(
    addressFields,
    (newAddressFields, prop, propName) => {
      const fieldRule = getField(propName, rules)

      newAddressFields[propName] = prop

      if (fieldRule && fieldRule.mask) {
        newAddressFields[propName] = {
          ...prop,
          ...(prop.value ? { value: msk(prop.value, fieldRule.mask) } : {}),
        }
      }

      return newAddressFields
    },
    {}
  )
}

export function addFocusToNextInvalidField(fields, rules) {
  const invalidFilledField = getFirstInvalidFilledField(fields, rules)

  if (invalidFilledField) {
    const { fieldName, field } = invalidFilledField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  const requiredField = getFirstRequiredFieldNotFilled(fields, rules)

  if (requiredField) {
    const { fieldName, field } = requiredField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  return fields
}

function getFirstInvalidFilledField(fields, rules) {
  const allFieldsVisited = addNewField(fields, 'visited', true)
  const validatedFields = validateChangedFields(
    allFieldsVisited,
    allFieldsVisited,
    rules
  )

  const orderedValidatedFields = orderFieldsByRules(validatedFields, rules)

  const firstInvalidField = find(
    orderedValidatedFields,
    field => field && field.valid === false
  )

  if (firstInvalidField) {
    return {
      fieldName: firstInvalidField.name,
      field: {
        ...validatedFields[firstInvalidField.name],
        focus: true,
      },
    }
  }

  return null
}

function orderFieldsByRules(fields, rules) {
  return reduce(
    fields,
    (acc, field, fieldName) => {
      const index = findIndex(
        rules.fields,
        ruleField => ruleField.name === fieldName
      )
      if (index === -1) return acc

      acc[index] = { ...field, name: fieldName }
      return acc
    },
    Array(Object.keys(fields).length)
  )
}

function getFirstRequiredFieldNotFilled(fields, rules) {
  const requiredFields = filter(rules.fields, field => field.required)
  const requiredFieldsNames = map(requiredFields, field => field.name)

  const fieldsNames = Object.keys(fields)
  const requiredFieldNotFilled = difference(requiredFieldsNames, fieldsNames)

  if (requiredFieldNotFilled && requiredFieldNotFilled.length > 0) {
    const nextRequiredFieldName = requiredFieldNotFilled[0]

    return { fieldName: nextRequiredFieldName, field: { focus: true } }
  }

  return null
}
