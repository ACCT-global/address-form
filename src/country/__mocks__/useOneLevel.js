import { ONE_LEVEL } from '../../constants'
import { firstLevelPostalCodes } from '../../transforms/postalCodes'
import {
  getOneLevel,
  getTwoLevels,
} from '../../transforms/addressFieldsOptions'

const countryData = {
  Azuay: {
    OÃ±a: '0000',
    Paute: '0000',
    Pucara: '0000',
  },
  Bolivar: {
    Caluma: '0001',
    Chillanes: '0001',
  },
}

export default {
  country: 'ECU',
  abbr: 'EC',
  postalCodeFrom: ONE_LEVEL,
  postalCodeLevel: 'state',
  firstLevelPostalCodes: firstLevelPostalCodes(countryData),
  fields: [
    {
      name: 'street',
      label: 'street',
      required: true,
      size: 'xlarge',
    },
    {
      name: 'number',
      label: 'number',
      required: true,
      size: 'mini',
    },
    {
      name: 'complement',
      label: 'complement',
      size: 'large',
    },
    {
      name: 'state',
      label: 'province',
      required: true,
      size: 'large',
      options: getOneLevel(countryData),
    },
    {
      name: 'city',
      label: 'city',
      required: true,
      size: 'large',
      basedOn: 'state',
      optionsMap: getTwoLevels(countryData),
    },
  ],
  geolocation: {
    postalCode: {
      valueIn: 'long_name',
      types: ['postal_code'],
      required: false,
      handler: address => {
        if (
          !address.state ||
          !address.city ||
          !address.state.value ||
          !address.city.value
        ) {
          return address
        }

        if (
          countryData[address.state.value] &&
          countryData[address.state.value][address.city.value]
        ) {
          address.postalCode = {
            value: countryData[address.state.value][address.city.value],
          }
        }

        return address
      },
    },
    number: { valueIn: 'long_name', types: ['street_number'], required: false },
    street: { valueIn: 'long_name', types: ['route'], required: false },
    state: {
      valueIn: 'long_name',
      types: ['administrative_area_level_1'],
      required: false,
    },
    city: {
      valueIn: 'long_name',
      types: ['administrative_area_level_2', 'locality'],
      required: false,
    },
  },
}
