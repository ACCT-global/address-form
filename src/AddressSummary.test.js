import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import AddressSummary from './AddressSummary'
import address from './__mocks__/addressWithoutValidation'
import usePostalCode from './country/__mocks__/usePostalCode'
import useOneLevel from './country/__mocks__/useOneLevel'
import displayBrazil from './country/__mocks__/displayBrazil'
import displayUSA from './country/__mocks__/displayUSA'

describe('AddressSummary', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(
      <AddressSummary address={address} rules={usePostalCode} />,
    )
    expect(wrapper.find('div')).toHaveLength(1)
  })

  it('should render address summary with all its data', () => {
    const tree = renderer.create(
      <AddressSummary
        address={{
          ...address,
          street: 'Praia de Botafogo',
          number: '300',
          complement: '2nd floor',
          neighborhood: 'Botafogo',
          city: 'Rio de Janeiro',
          state: 'AM',
          country: 'BRA',
        }}
        rules={usePostalCode}
      />,
    )

    expect(tree).toMatchSnapshot()
  })

  it('should render the postal code if country show its input', () => {
    const wrapper = shallow(
      <AddressSummary address={address} rules={usePostalCode} />,
    )

    expect(wrapper.find('.postal-code')).toHaveLength(1)
  })

  it('should not render postal code if the country sets it by another input', () => {
    const wrapper = shallow(
      <AddressSummary address={address} rules={useOneLevel} />,
    )

    expect(wrapper.find('.postal-code')).toHaveLength(0)
  })

  it('should render child component', () => {
    function MyChild() {
      return <div />
    }

    const wrapper = shallow(
      <AddressSummary address={address} rules={useOneLevel}>
        <MyChild />
      </AddressSummary>,
    )

    expect(wrapper.find('MyChild')).toHaveLength(1)
  })

  it('should call click handler', () => {
    const handleClick = jest.fn()

    const wrapper = shallow(
      <AddressSummary
        address={address}
        rules={useOneLevel}
        canEditData={false}
        onClickMaskedInfoIcon={handleClick}
      />,
    )

    const maskedInfoIcon = wrapper.find('.client-masked-info')
    maskedInfoIcon.simulate('click', { preventDefault() {} })

    expect(handleClick).toHaveBeenCalled()
  })

  it('should render gift list address', () => {
    const tree = renderer.create(
      <AddressSummary
        giftRegistryDescription={'JoÃ£o da Silva'}
        address={address}
        rules={usePostalCode}
      />,
    )

    expect(tree).toMatchSnapshot()
  })

  it('should format address according to rules', () => {
    const brazilianAddress = renderer.create(
      <AddressSummary
        address={{
          ...address,
          street: 'Av. Praia de Botafogo',
          number: '300',
          complement: 'ap. 322',
          neighborhood: 'Botafogo',
          city: 'Rio de Janeiro',
          state: 'RJ',
          country: 'BRA',
        }}
        rules={displayBrazil}
      />,
    )

    const americanAddress = renderer.create(
      <AddressSummary
        address={{
          ...address,
          street: '1 Infinite Loop',
          number: null,
          complement: 'Suite 306',
          neighborhood: null,
          city: 'Cupertino',
          state: 'CA',
          country: 'USA',
        }}
        rules={displayBrazil}
      />,
    )

    expect(brazilianAddress).toMatchSnapshot()
    expect(americanAddress).toMatchSnapshot()
  })

  it('should format address according to rules', () => {
    const brazilianAddress = renderer.create(
      <AddressSummary
        address={{
          ...address,
          street: 'Av. Praia de Botafogo',
          number: '300',
          complement: 'ap. 322',
          neighborhood: 'Botafogo',
          city: 'Rio de Janeiro',
          state: 'RJ',
          country: 'BRA',
          postalCode: '22250-040',
        }}
        rules={displayBrazil}
      />,
    )

    const americanAddress = renderer.create(
      <AddressSummary
        address={{
          ...address,
          street: '1 Infinite Loop',
          number: null,
          complement: 'Suite 306',
          neighborhood: null,
          city: 'Cupertino',
          state: 'CA',
          country: 'USA',
          postalCode: '95014',
        }}
        rules={displayUSA}
      />,
    )

    expect(brazilianAddress).toMatchSnapshot()
    expect(americanAddress).toMatchSnapshot()
  })
})
