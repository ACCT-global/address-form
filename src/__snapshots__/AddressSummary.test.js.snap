// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressSummary should format address according to rules 1`] = `
<div
  className="address-summary"
>
  <span
    className="street"
  >
    Av. Praia de Botafogo
  </span>
  <span
    className="number-comma"
  >
    ,
  <span
    className="number"
  >
    300
  </span>
  <span
    className="complement-comma"
  >
    ,
  </span>
  <span
    className="complement"
  >
    ap. 322
  </span>
  <br />
  <span
    className="neighborhood"
  >
    Botafogo
  </span>
  <span
    className="city-dash"
  >
     -
  </span>
  <span
    className="city"
  >
    Rio de Janeiro
  </span>
  <span
    className="state-dash"
  >
     -
  </span>
  <span
    className="state"
  >
    RJ
  </span>
  <br />
  <span
    className="postal-code"
  >
    22250-040
  </span>
  <br />
  <span
    className="country"
  >
    Brasil
  </span>
</div>
`;

exports[`AddressSummary should format address according to rules 2`] = `
<div
  className="address-summary"
>
  <span
    className="street"
  >
    1 Infinite Loop
  </span>
  <span
    className="complement-comma"
  >
    ,
  </span>
  <span
    className="complement"
  >
    Suite 306
  </span>
  <br />
  <span
    className="city"
  >
    Cupertino
  </span>
  <span
    className="state-comma"
  >
     ,
  </span>
  <span
    className="state"
  >
    CA
  </span>
  <span
    className="postal-code"
  >
    95014
  </span>
  <span
    className="country"
  >
    USA
  </span>
</div>
`;

exports[`AddressSummary should render address summary with all its data 1`] = `
<div
  className="address-summary"
>
  <span
    className="street"
  >
    Praia de Botafogo
  </span>

  <span
    className="number"
  >
    300
  </span>
  <span
    className="complement-comma"
  >
    ,
  </span>
  <span
    className="complement"
  >
    2nd floor
  </span>
  <br />
  <span
    className="neighborhood"
  >
    Botafogo
  </span>
  <span
    className="city-dash"
  >
     -
  </span>
  <span
    className="city"
  >
    Rio de Janeiro
  </span>
  <span
    className="state-dash"
  >
     -
  </span>
  <span
    className="state"
  >
    AM
  </span>
  <br />
  <span
    className="postal-code"
  />
  <span
    className="country-dash"
  >
     -
  </span>
  <span
    className="country"
  >
    BRA
  </span>
</div>
`;

exports[`AddressSummary should render gift list address 1`] = `
<span>
  <span>
    At address of:
  </span>

  <strong
    className="gift-list-name"
  >
    Jo√£o da Silva
  </strong>
</span>
`;
